linear_operator/operators/diag_linear_operator.py

```diff
@@ -44,92 +48,115 @@ def _bilinear_derivative(self, left_vecs: Tensor, right_vecs: Tensor) -> Tuple[T
+        upper: Optional[bool] = False,
+    ) -> Union[Float[LinearOperator, "... N M"], Float[Tensor, "... N M"]]:
         return rhs / self._diag.unsqueeze(-1).pow(2)
 
-    def _expand_batch(self, batch_shape: torch.Size) -> "DiagLinearOperator":
+    def _expand_batch(
+        self: Float[LinearOperator, "... M N"], batch_shape: Union[torch.Size, List[int]]
+    ) -> Float[LinearOperator, "... M N"]:
         return self.__class__(self._diag.expand(*batch_shape, self._diag.size(-1)))
 
-    def _diagonal(self) -> Tensor:
+    def _diagonal(self: Float[LinearOperator, "... M N"]) -> Float[torch.Tensor, "... N"]:
         return self._diag
 
-    def _get_indices(
-        self,
-        row_index: torch.LongTensor,
-        col_index: torch.LongTensor,
-        *batch_indices: Tuple[torch.LongTensor, ...],
-    ) -> Tensor:
+    def _get_indices(self, row_index: IndexType, col_index: IndexType, *batch_indices: IndexType) -> torch.Tensor:
         res = self._diag[(*batch_indices, row_index)]
         # If row and col index don't agree, then we have off diagonal elements
         # Those should be zero'd out
         res = res * torch.eq(row_index, col_index).to(device=res.device, dtype=res.dtype)
         return res
 
-    def _mul_constant(self, constant: Tensor) -> "DiagLinearOperator":
-        return self.__class__(self._diag * constant.unsqueeze(-1))
+    def _mul_constant(
+        self: Float[LinearOperator, "*batch M N"], other: Union[float, torch.Tensor]
+    ) -> Float[LinearOperator, "*batch M N"]:
+        return self.__class__(self._diag * other.unsqueeze(-1))
 
-    def _mul_matrix(self, other: Tensor) -> "DiagLinearOperator":
+    def _mul_matrix(
+        self: Float[LinearOperator, "... #M #N"],
+        other: Union[Float[torch.Tensor, "... #M #N"], Float[LinearOperator, "... #M #N"]],
+    ) -> Float[LinearOperator, "... M N"]:
         return DiagLinearOperator(self._diag * other._diagonal())
 
-    def _prod_batch(self, dim: int) -> "DiagLinearOperator":
+    def _prod_batch(self, dim: int) -> LinearOperator:
         return self.__class__(self._diag.prod(dim))
 
-    def _root_decomposition(self) -> "DiagLinearOperator":
+    def _root_decomposition(
+        self: Float[LinearOperator, "... N N"]
+    ) -> Union[Float[torch.Tensor, "... N N"], Float[LinearOperator, "... N N"]]:
         return self.sqrt()
 
-    def _root_inv_decomposition(self, initial_vectors: Optional[Tensor] = None) -> "DiagLinearOperator":
+    def _root_inv_decomposition(
+        self: Float[LinearOperator, "*batch N N"],
+        initial_vectors: Optional[torch.Tensor] = None,
+        test_vectors: Optional[torch.Tensor] = None,
+    ) -> Union[Float[LinearOperator, "... N N"], Float[Tensor, "... N N"]]:
         return self.inverse().sqrt()
 
     def _size(self) -> torch.Size:
-        return self._diag.shape + self._diag.shape[-1:]
+        return torch.Size([*self._diag.shape, self._diag.shape[-1]])
```

Comment:
This change broke the case where `self._diag` is a scalar tensor. Will put up a fix